cli/src/config.rs

```diff
@@ -59,7 +60,7 @@ pub struct AnnotatedValue {
 /// 2. Base environment variables
 /// 3. [User config](https://github.com/martinvonz/jj/blob/main/docs/config.md#configuration)
 /// 4. Repo config `.jj/repo/config.toml`
-/// 5. TODO: Workspace config `.jj/config.toml`
+/// 5. Workspace config `.jj/config.toml`
```

Comment:
I'm wondering whether `.jj/working-copy/config.toml` might be a better location for this. Otherwise, this emphasizes the working-copy config over the repo-level config, and I'm not sure that's desired.

Reply:
> I'm wondering whether `.jj/working_copy/config.toml` might be a better location for this.

My feeling is that `.jj/working_copy` is the store directory dedicated for the `WorkingCopy` implementation, so better to not add arbitrary files in it.

Reply:
OK.

I suppose, if need be, we could later move `.jj/working_copy` to `.jj/workspace/working_copy` and then `.jj/config.toml` to `.jj/workspace/config.toml`.

Reply:
I'm still a bit worried about people thinking that workspaces are broken if they end up using `.jj/config.toml` instead of `.jj/repo/config.toml`.

WDYT of putting this file in `.jj/workspace/config.toml` now, and then moving `working_copy` to that dir later?

This is made better by `jj config edit` and `jj config path` existing, so I don't think it's absolutely critical, but it would be nice to not have to move the config file later (with all the doc changes and backwards compatibility work that involves)

Reply:
Maybe the file can be renamed to `.jj/workspace-config.toml`, then?

BTW, do we have any practical use case of workspace-specific configuration? If we don't, and if the existence of `.jj/config.toml` is confusing, it might be better to not support workspace config at all. Not having workspace config might help simplify `IncludeIf`-like conditional. #616

Reply:
> Maybe the file can be renamed to .jj/workspace-config.toml, then?

Do you prefer this to `.jj/workspace/config.toml`? Another thing to consider is that at some point we might want a per-workspace ignore file (that might be more important than a per-workspace config, though it might also be more difficult to sync with Git) that could be called `.jj/workspace/.gitignore` or `.jj/workspace/exclude`.

If we do prefer `.jj/workspace-config.toml`, I think it would solve the confusion problem. Though, should we then also have a `.jj/repo-config.toml`?

> BTW, do we have any practical use case of workspace-specific configuration?

It's a fair question. I can imagine, for example, making a sparse workspace and turning off fs-monitor there, but I haven't immediately thought of a very compelling use-case. But that might be because I don't use workspaces very much at the moment.

Reply:
> > Maybe the file can be renamed to .jj/workspace-config.toml, then?
> 
> Do you prefer this to `.jj/workspace/config.toml`?

Yes. It mirrors the internal data model (workspace -> repo), and I like somewhat flattened structure for user-facing data. `.jj/repo/config.toml` isn't neat, but repo config has to be naturally accessible from multiple workspaces.

Reply:
I keep coming back to this occasionally, and I will say that while your argument makes some sense, I still feel that `.jj/workspace/config.toml` is the clearest option given that `.jj/repo/config.toml` exists. Putting a file in the top-level `.jj` dir will encourage people to pay attention to it, whereas they should generally pay more attention to `.jj/repo/config.toml`.

That said, I did suggest some other options above that you seemed happier with, and that would also work. It also depends a lot on the use-case. Currently, the use-case for the workspace config that I have in mind (and the reason I came back to this discussion) is to use it for setting up ignored/untracked files, as in https://github.com/martinvonz/jj/pull/4338#discussion_r1741304289 (which is one of several possible approaches).

Reply:
I agree with @yuja that `.jj/config.toml` most closely matches the model. I see `.jj/repo/` as containing repo state, `.jj/working_copy/`as containing working-copy state, and `.jj/` as being the whole workspace state (the combination of repo and working-copy state).

> Putting a file in the top-level `.jj` dir will encourage people to pay attention to it, whereas they should generally pay more attention to `.jj/repo/config.toml`.

That's a good point. Perhaps we can help avoid that by writing a comment to the top of the file telling the user that it's specific to the workspace and that they might be looking for the repo-level config.

Reply:
>  Perhaps we can help avoid that by writing a comment to the top of the file

I would prefer to do a bit more. If we don't want a `.jj/workspace/` dir, I'd rename the file to something like `.jj/workspace-config.toml`. Just a comment on its own feels a bit insufficient to me[^fluff], though if people prefer it, it wouldn't be the end of the world.

[^fluff]: I'm imagining one person telling what to do to another slightly confused and rushed person over the phone. If they forget to stress this point, the other person might not read the comment because they are following instructions. I want to give them some reason to ask "hey, it's not called config, it's called workspace-config, is that OK?" or to make it likely they find both files.

I'm starting to suspect we use the term "workspace" very slightly differently. As a user I don't think of "workspace state" (in the sense of what the `jj workspace` command manages) as including the repo state. You are right that in the implementation, it does.

To be clear, my intention was to keep the workspace `config.toml` next to `working_copy`, and move both over to `.jj/workspace`. But if I haven't convinced you about the `workspace` directory at this point, we should probably just go with the other option.

Reply:
Calling the file `.jj/workspace-config.toml` works for me. Are you also okay with that, @yuja? Sorry that this got stalled, @HybridEidolon.

Reply:
> Calling the file `.jj/workspace-config.toml` works for me.

Yes, that seems okay.

Another option is to not add workspace config until we find a real use case.

> BTW, do we have any practical use case of workspace-specific configuration? If we don't, and if the existence of `.jj/config.toml` is confusing, it might be better to not support workspace config at all. Not having workspace config might help simplify `IncludeIf`-like conditional. #616

Reply:
I was thinking that the config for not ignoring the executable bit in the working copy (#4478) could be stored in a workspace-level config instead of detecting support every time we try to snapshot the working copy.

Reply:
> I was thinking that the config for not ignoring the executable bit in the working copy (#4478) could be stored in a workspace-level config

Yes, but I assumed it's practically okay to set it to repo/config.toml. (Maybe I should say "immediate need", not "real use case".)

Reply:
If we go in the direction of supporting "IncludeIf" and conditional config, it might make sense to put workspace config inside repo config under a key like `[if.workspace.default]` (assuming `if.something` is how other conditional config can be setup; I'd want these to have similar syntax if possible, thought it's not absolutely necessary).

For something like executable bits, we could eventually have a fancier configs like "Config for all workspaces under `/mnt/c`" (or whatever the WSL mount for the NTFS C: drive is).

At the same time, I'm not sure whether the conditional config idea is more practical for now (or in general). It might be easier to have `workspace-config.toml` for now (or even permanently).

Reply:
For context, the reason I went ahead and implemented the workspace config is because I had a complex repository scenario where multiple related-but-divergent trunks shared a lot of history, and wanted to use multiple workspaces to manage the different related codebases with different jj templates and revsets for each.

Reply:
Ok, if there's a real need, we shouldn't be too nitpicky. Let's land `.jj/config.toml` or `.jj/workspace-config.toml`. I think @ilyagr is okay with `.jj/workspace-config.toml`.

Reply:
I'm OK with `.jj/workspace-config.toml`. I appreciate Eidolon explaining one more use-case; I don't have a clear opinion on what the conclusion from that or from Martin's use-case should be.